FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,80 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				if let existingCache = try ManagedCache.fetch(in: context) {
+					completion(.found(feed: existingCache.feed.toLocal(), timestamp: existingCache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let existingCache = try ManagedCache.fetch(in: context)
+				existingCache.map(context.delete)
+
+				let newCache = ManagedCache(context: context)
+				newCache.timestamp = timestamp
+				newCache.feed = feed.toManaged(context: context)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let existingCache = try ManagedCache.fetch(in: context)
+				existingCache.map(context.delete)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+}
+
+private extension NSOrderedSet {
+	func toLocal() -> [LocalFeedImage] {
+		array
+			.compactMap { $0 as? ManagedFeedImage }
+			.map { managedFeedImage in
+				return LocalFeedImage(
+					id: managedFeedImage.id,
+					description: managedFeedImage.imageDescription,
+					location: managedFeedImage.location,
+					url: managedFeedImage.url
+				)
+			}
+	}
+}
+
+private extension Array where Element == LocalFeedImage {
+	func toManaged(context: NSManagedObjectContext) -> NSOrderedSet {
```

Comment:
This function name can be confusing, since it's being called with `feed.toManaged(context:`, while actually returning a `NSOrderedSet` of `ManagedFeedImage`s.
A suggestion would be to move this logic as a static method in `ManagedFeedImage`, and it would read as:
```swift
ManagedFeedImage.images(from: feed, in: context)
```
Another option would be renaming it so its called as:
```swift
feed.managedImages(in: context)
```
These are suggestions to improve this function naming and clarify what type of object it returns.
```suggestion
	func managedImages(in context: NSManagedObjectContext) -> NSOrderedSet {
```